### DISCUSSION TITLE
Timing Sagas

### DISCUSSION THREAD

**Discussion by sdb1228** [redux-saga/redux-saga#2253] [Q&A]
Hello! I am wondering if there some middleware someone is aware of that would allow me to track an individual sagas from start to finish? I have been experimenting with sagamonitor middleware but that logs every single effect and you can have multiple effects inside a saga.  Example:

``` 
function* randomSagaFunction(action) {
  ...does stuff and has multiple effects
}
export function* onRandomFunction() {
    yield takeEvery(ACTION, randomSagaFunction);
}
```

Then in my application saga I fork my sagas like 
```
  yield all(
        [
            ...randomSagas,
        ].map(fork)
```

I would ultimately like to track how long randomSagaFunction takes without adding instrumentation to every single saga function in my app.  Wondering if there is a middleware that would accomplish this that I am missing....

**Comment by neurosnap** [comment]
Using something like [saga-query](https://github.com/neurosnap/saga-query) could help

```ts
import { createPipe } from 'saga-query';

// saga middleware system
export const timer = createPipe();
timer.use(function* (ctx, next) {
  const t0 = performance.now();
  yield next();
  const t1 = performance.now();
  console.log(`Call to ${ctx.name} took ${t1 - t0} milliseconds.`);
});
timer.use(timer.routes());

// your saga function
function* randomSagaFunction(action) {
  ...does stuff and has multiple effects
}

const randomSaga = timer.create('random-saga', function* (ctx) {
  yield call(randomSagaFunction, ctx.action);
})

// use it in another saga
function* someOtherSaga() {
  yield call(randomSaga.run, randomSaga())
}
// call it directly
store.dispatch(randomSaga());
```

It seems a little round-about but once you have `saga-query` installed into your system, I think you'll find using it easier than hand-crafting all of your sagas.  For example, it handles combining all of your sagas into one forked and fault-tolerant saga using [redux-saga-creator](https://github.com/neurosnap/redux-saga-creator).  It also allows for a wide variety of ways to improve flow control using a koa-like middleware system.

**Comment by neurosnap** [comment]
I was inspired by this post to make `saga-query` easier to use for this use case, so I added a couple of functions to make this easier:

```ts
import { performanceMonitor, createPipe, wrap, PerfCtx, delay } from 'saga-query';

const thunks = createPipe<PerfCtx>();
thunks.use(function* (ctx, next) {
  yield next();
  console.log(`calling ${ctx.name} took ${ctx.performance} ms`);
});
thunks.use(performanceMonitor);
thunks.use(thunks.routes());

// your saga function
function* randomSagaFunction(action) {
  ...does stuff and has multiple effects
}

const slow = thunks.create('something-slow', wrap(randomSagaFunction));

store.dispatch(slow());
// calling something-slow took 10000 ms
```

`performanceMonitor` attaches a property to the `ctx` object that logs the time it takes to run a saga and `wrap` just makes it easier to call a saga without the boilerplate.

happy to read any thoughts!
